source/dhtslib/vcf/header.d

```diff
@@ -61,4 +318,415 @@ struct VCFHeader
     pragma(inline, true)
     @property int nsamples() { return bcf_hdr_nsamples(this.hdr); }
 
+    int getSampleId(string sam){
+        auto ret = bcf_hdr_id2int(this.hdr, HeaderDictTypes.SAMPLE, toUTFz!(char *)(sam));
+        if(ret == -1) hts_log_error(__FUNCTION__, "Couldn't find sample in header: " ~ sam);
+        return ret;
+    }
+
+    // TODO
+    /// copy header lines from a template without overwiting existing lines
+    void copyHeaderLines(bcf_hdr_t *other)
+    {
+        assert(this.hdr != null);
+        assert(0);
+        //    bcf_hdr_t *bcf_hdr_merge(bcf_hdr_t *dst, const(bcf_hdr_t) *src);
+    }
+
+    /// Add sample to this VCF
+    /// * int bcf_hdr_add_sample(bcf_hdr_t *hdr, const(char) *sample);
+    int addSample(string name)
+    in { assert(name != ""); }
+    do
+    {
+        assert(this.hdr != null);
+
+        bcf_hdr_add_sample(this.hdr, toStringz(name));
+
+        // AARRRRGGGHHHH
+        // https://github.com/samtools/htslib/issues/767
+        bcf_hdr_sync(this.hdr);
+
+        return 0;
+    }
+
+    /// Add a new header line
+    int addHeaderLineKV(string key, string value)
+    {
+        // TODO check that key is not INFO, FILTER, FORMAT (or contig?)
+        string line = format("##%s=%s", key, value);
+
+        return bcf_hdr_append(this.hdr, toStringz(line));

```

Comment:
probably need `bcf_hdr_sync` call here (see `addHeaderLineRaw` ; also check `bcf_hdr_append` to make sure it doesn't call it -- `bcf_hdr_add_sample` doesn't, see my github issue above)

An alternative which could be faster in some rare circum stances, although unlikely not as I point out in the github issue above, would be to FOREGO the sync in all of these update functions , add a `dirty` flag to the struct overall , which could be read directly from the underlying record (htslib uses a dirty flag) ,a nd then check for dirty in any READ methods, and sync if necessary.

But probably it's just easier to add a sync here since we're alreadying syncing elsewhere in other writer fns

Reply:
Yeah I tried to add that anywhere I was altering a header line or adding a new one. Will add.

Reply:
addressed in f2b29df